File: packages/rust/src/spec.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct EnvironmentSpec {
    pub adapter_id: String,
    pub env_overrides: HashMap<String, String>,
    pub seed_files: HashMap<String, String>,
    pub file_modes: HashMap<String, u32>,
    pub credentials: HashMap<String, String>,
    pub prefix: String,
}

impl Default for EnvironmentSpec {
    fn default() -> Self {
        Self {
            adapter_id: "generic".into(),
            env_overrides: HashMap::new(),
            seed_files: HashMap::new(),
            file_modes: HashMap::new(),
            credentials: HashMap::new(),
            prefix: "agent-venv-".into(),
        }
    }
}

impl EnvironmentSpec {
    pub fn builder() -> EnvironmentSpecBuilder {
        EnvironmentSpecBuilder::default()
    }
}

#[derive(Default)]
pub struct EnvironmentSpecBuilder {
    spec: EnvironmentSpec,
}

impl EnvironmentSpecBuilder {
    pub fn adapter_id(mut self, id: impl Into<String>) -> Self {
        self.spec.adapter_id = id.into();
        self
    }
    pub fn env_override(mut self, k: impl Into<String>, v: impl Into<String>) -> Self {
        self.spec.env_overrides.insert(k.into(), v.into());
        self
    }
    pub fn seed_file(mut self, rel: impl Into<String>, content: impl Into<String>) -> Self {
        self.spec.seed_files.insert(rel.into(), content.into());
        self
    }
    pub fn credential(mut self, rel: impl Into<String>, content: impl Into<String>) -> Self {
        self.spec.credentials.insert(rel.into(), content.into());
        self
    }
    pub fn file_mode(mut self, rel: impl Into<String>, mode: u32) -> Self {
        self.spec.file_modes.insert(rel.into(), mode);
        self
    }
    pub fn prefix(mut self, prefix: impl Into<String>) -> Self {
        self.spec.prefix = prefix.into();
        self
    }
    pub fn build(self) -> EnvironmentSpec {
        self.spec
    }
}

File: packages/go/spec.go
package agentvenv

import (
	"io/fs"
	"path/filepath"
	"strings"
)

// EnvironmentSpec describes a profile directory to materialize.
// See spec/environment-spec.schema.json.
type EnvironmentSpec struct {
	AdapterID    string                 // "claude-code" / "codex" / "generic"
	EnvOverrides map[string]string      // values may contain "$EPHEMERAL_HOME"
	SeedFiles    map[string]string      // relative path -> contents
	FileModes    map[string]fs.FileMode // relative path -> mode bits (Unix)
	Credentials  map[string]string      // relative path -> contents (default mode 0600)
	Prefix       string                 // ephemeral tmpdir prefix
}

func (s *EnvironmentSpec) adapterIDOrDefault() string {
	if s.AdapterID == "" {
		return "generic"
	}
	return s.AdapterID
}

func (s *EnvironmentSpec) prefixOrDefault() string {
	if s.Prefix == "" {
		return "agent-venv-"
	}
	return s.Prefix
}

func (s *EnvironmentSpec) validate() error {
	for rel := range s.SeedFiles {
		if err := validateRelPath(rel); err != nil {
			return err
		}
	}
	for rel := range s.Credentials {
		if err := validateRelPath(rel); err != nil {
			return err
		}
	}
	for rel := range s.FileModes {
		if err := validateRelPath(rel); err != nil {
			return err
		}
	}
	return nil
}

func validateRelPath(rel string) error {
	if rel == "" {
		return newErr(ErrInvalidEnvironmentSpec, "empty path", nil)
	}
	if filepath.IsAbs(rel) {
		return newErr(ErrInvalidEnvironmentSpec, "absolute path: "+rel, nil)
	}
	cleaned := filepath.Clean(rel)
	if cleaned == ".." || strings.HasPrefix(cleaned, ".."+string(filepath.Separator)) {
		return newErr(ErrInvalidEnvironmentSpec, "path escapes profile: "+rel, nil)
	}
	return nil
}
